nibabel/streamlines/compact_list.py

```diff
@@ -0,0 +1,232 @@
+        # to use less memory. For example, if the compact list being copied
+        # is the result of a slicing operation on a compact list.
+        clist = CompactList()
+        total_lengths = np.sum(self._lengths)
+        clist._data = np.empty((total_lengths,) + self._data.shape[1:],
+                               dtype=self._data.dtype)
+
+        next_offset = 0
+        offsets = []
+        for offset, length in zip(self._offsets, self._lengths):
+            offsets.append(next_offset)
+            clist._data[next_offset:next_offset+length] = self._data[offset:offset+length]
+            next_offset += length
+
+        clist._offsets = np.asarray(offsets)
+        clist._lengths = self._lengths.copy()
+
+        return clist
+
+    def __getitem__(self, idx):
+        """ Gets element(s) through indexing.
+
+        Parameters
+        ----------
+        idx : int, slice or list
+            Index of the element(s) to get.
+
+        Returns
+        -------
+        ndarray object(s)
+            When `idx` is an int, returns a single ndarray.
+            When `idx` is either a slice or a list, returns a list of ndarrays.
+        """
+        if isinstance(idx, int) or isinstance(idx, np.integer):
+            start = self._offsets[idx]
+            return self._data[start:start+self._lengths[idx]]
+
+        elif isinstance(idx, slice) or isinstance(idx, list):
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = self._offsets[idx]
+            clist._lengths = self._lengths[idx]
+            return clist
+
+        elif isinstance(idx, np.ndarray) and np.issubdtype(idx.dtype, np.integer):
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = self._offsets[idx]
+            clist._lengths = self._lengths[idx]
+            return clist
+
+        elif isinstance(idx, np.ndarray) and idx.dtype == np.bool:
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = [self._offsets[i]
+                              for i, take_it in enumerate(idx) if take_it]
+            clist._lengths = [self._lengths[i]
+                              for i, take_it in enumerate(idx) if take_it]
+            return clist
+
+        raise TypeError("Index must be either an int, a slice, a list of int"
+                        " or a ndarray of bool! Not " + str(type(idx)))
+
+    def __iter__(self):
+        if len(self._lengths) != len(self._offsets):
+            raise ValueError("CompactList object corrupted:"
+                             " len(self._lengths) != len(self._offsets)")
+
+        for offset, lengths in zip(self._offsets, self._lengths):
+            yield self._data[offset: offset+lengths]
+
+    def __len__(self):
+        return len(self._offsets)
+
+    def __repr__(self):
+        return repr(list(self))
+
+
+def save_compact_list(filename, clist):
```

Comment:
This function uses private attributes of CompactList - maybe better as a method of CompactList.